contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol

```diff
@@ -0,0 +1,1803 @@
+  function _cancelTransferTo(address _to, uint256 _amount) internal virtual {
+    _to; // Prevents unused var compiler and natspec complaints.
+    _amount; // Prevents unused var compiler and natspec complaints.
+  }
+
+  /**
+    @notice
+    Holders can redeem their tokens to claim the project's overflowed tokens, or to trigger rules determined by the project's current funding cycle's data source.
+
+    @dev
+    Only a token holder or a designated operator can redeem its tokens.
+
+    @param _holder The account to redeem tokens for.
+    @param _projectId The ID of the project to which the tokens being redeemed belong.
+    @param _tokenCount The number of project tokens to redeem, as a fixed point number with 18 decimals.
+    @param _minReturnedTokens The minimum amount of terminal tokens expected in return, as a fixed point number with the same amount of decimals as the terminal.
+    @param _beneficiary The address to send the terminal tokens to.
+    @param _memo A memo to pass along to the emitted event.
+    @param _metadata Bytes to send along to the data source, delegate, and emitted event, if provided.
+
+    @return reclaimAmount The amount of terminal tokens that the project tokens were redeemed for, as a fixed point number with 18 decimals.
+  */
+  function _redeemTokensOf(
+    address _holder,
+    uint256 _projectId,
+    uint256 _tokenCount,
+    uint256 _minReturnedTokens,
+    address payable _beneficiary,
+    string memory _memo,
+    bytes memory _metadata
+  ) internal returns (uint256 reclaimAmount) {
+    // Can't send reclaimed funds to the zero address.
+    if (_beneficiary == address(0)) revert REDEEM_TO_ZERO_ADDRESS();
+
+    // Define variables that will be needed outside the scoped section below.
+    // Keep a reference to the funding cycle during which the redemption is being made.
+    JBFundingCycle memory _fundingCycle;
+
+    // Keep a reference to the amount being reclaimed that should have fees withheld from.
+    uint256 _feeEligibleDistributionAmount;
+
+    // Keep a reference to the amount of discount to apply to the fee.
+    uint256 _feeDiscount;
+
+    // Scoped section prevents stack too deep. `_delegateAllocations` only used within scope.
+    {
+      JBRedemptionDelegateAllocation3_1_1[] memory _delegateAllocations;
+
+      // Record the redemption.
+      (
+        _fundingCycle,
+        reclaimAmount,
+        _delegateAllocations,
+        _memo
+      ) = IJBSingleTokenPaymentTerminalStore3_1_1(store).recordRedemptionFor(
+        _holder,
+        _projectId,
+        _tokenCount,
+        _memo,
+        _metadata
+      );
+
+      // Set the reference to the fee discount to apply. No fee if the beneficiary is feeless or if the redemption rate is at its max.
+      _feeDiscount = isFeelessAddress[_beneficiary] ||
+        (_fundingCycle.redemptionRate() == JBConstants.MAX_REDEMPTION_RATE &&
+          _fundingCycle.ballotRedemptionRate() == JBConstants.MAX_REDEMPTION_RATE) ||
```

Comment:
Shouldn't this check the ballot is Active (eg `fundingCycleStore.currentBallotStateOf(_projectId);`)?

Reply:
it could, but thats a whole nother lookup. the local contract doesn't have a fundingCycleStore instance. im down to go this route, or just keep it so fees apply if either setting isn't 100%.